pkg/processor/procbuiltin/unwrap.go

```diff
@@ -0,0 +1,310 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package procbuiltin
+
+import (
+	"context"
+	"encoding/json"
+
+	"github.com/conduitio/conduit/pkg/foundation/cerrors"
+	"github.com/conduitio/conduit/pkg/foundation/log"
+	"github.com/conduitio/conduit/pkg/foundation/multierror"
+	"github.com/conduitio/conduit/pkg/inspector"
+	"github.com/conduitio/conduit/pkg/processor"
+	"github.com/conduitio/conduit/pkg/record"
+	"github.com/rs/zerolog"
+)
+
+type unwrapProcessor struct {
+	unWrapper unWrapper
+
+	inInsp  *inspector.Inspector
+	outInsp *inspector.Inspector
+}
+
+// unWrapper unwraps the formatted record from the openCDC record
+type unWrapper interface {
+	// Unwrap gets the unwrapped record
+	Unwrap(record.Record) (record.Record, error)
+}
+
+const (
+	unwrapPayloadProcType = "unwrap"
+
+	unwrapConfigFormat = "format"
+
+	FormatDebezium     = "debezium"
+	FormatKafkaConnect = "kafka-connect"
+)
+
+func init() {
+	processor.GlobalBuilderRegistry.MustRegister(unwrapPayloadProcType, UnwrapBuilder)
+}
+
+func UnwrapBuilder(config processor.Config) (processor.Interface, error) {
+	if _, ok := config.Settings[unwrapConfigFormat]; !ok {
+		return nil, cerrors.Errorf("%s: %q config not specified", unwrapPayloadProcType, unwrapConfigFormat)
+	}
+	format := config.Settings[unwrapConfigFormat]
+	if format != FormatDebezium && format != FormatKafkaConnect {
+		return nil, cerrors.Errorf("%s: %q is not a valid format", unwrapPayloadProcType, format)
+	}
+
+	cw := zerolog.NewConsoleWriter()
+	cw.TimeFormat = "2006-01-02T15:04:05+00:00"
+	logger := zerolog.New(cw).With().Timestamp().Logger()
+	proc := &unwrapProcessor{
+		inInsp:  inspector.New(log.New(logger), inspector.DefaultBufferSize),
+		outInsp: inspector.New(log.New(logger), inspector.DefaultBufferSize),
+	}
+	switch format {
+	case FormatDebezium:
+		proc.unWrapper = &DebeziumUnWrapper{}
+	case FormatKafkaConnect:
+		proc.unWrapper = &KafkaConnectUnWrapper{}
+	}
+	return proc, nil
+}
+
+func (p *unwrapProcessor) Process(ctx context.Context, in record.Record) (record.Record, error) {
+	p.inInsp.Send(ctx, in)
+	out, err := p.unWrapper.Unwrap(in)
+	if err != nil {
+		return record.Record{}, cerrors.Errorf("%s: error unwrapping record: %w", unwrapPayloadProcType, err)
+	}
+	p.outInsp.Send(ctx, out)
+	return out, nil
+}
+
+func (p *unwrapProcessor) InspectIn(ctx context.Context) *inspector.Session {
+	return p.inInsp.NewSession(ctx)
+}
+
+func (p *unwrapProcessor) InspectOut(ctx context.Context) *inspector.Session {
+	return p.outInsp.NewSession(ctx)
+}
+
+/*
+	 example of kafka-connect record:
+		`{
+		 "payload": {
+		     "description": "desc",
+		     "id": 20
+		   },
+		 "schema": {} // will be ignored
+		}`
+*/
+type KafkaConnectUnWrapper struct{}
```

Comment:
❕ Thinking out loud - should we be more strict in this unwrapper and check that the record only contains 2 fields, `payload` and `schema`? Otherwise some data might get thrown away without the user knowing. I wonder if Kafka Connect allows records to have fields beside `payload` and `schema` (e.g. `{"payload":{...},"schema":{...},"foo":"bar"}` 🤔

Reply:
Let's not block because of this and not be too strict for no reason 😉 Resolving this comment.